pkg/cluster/agent_customization.go

```diff
@@ -0,0 +1,98 @@
+package cluster
+
+import (
+	"encoding/json"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// GetClusterAgentTolerations returns additional tolerations for the cluster agent if they have been user defined. If
+// not, nil is returned.
+func GetClusterAgentTolerations(cluster *v3.Cluster) []corev1.Toleration {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.AppendTolerations != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.AppendTolerations
+	}
+
+	return nil
+}
+
+// GetClusterAgentAffinity returns node affinity for the cluster agent if it has been user defined. If not, then the
+// default affinity is returned.
+func GetClusterAgentAffinity(cluster *v3.Cluster) *corev1.Affinity {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity
+	}
+
+	affinity, err := unmarshalAffinity(settings.ClusterAgentDefaultAffinity.Get())
+	if err != nil {
+		return settings.GetClusterAgentDefaultAffinity()
+	}
+
+	return affinity
+}
+
+// GetClusterAgentResourceRequirements returns resource requirements (cpu, memory) for the cluster agent if it has been
+// user defined. If not, nil is returned.
+func GetClusterAgentResourceRequirements(cluster *v3.Cluster) *corev1.ResourceRequirements {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.OverrideResourceRequirements != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.OverrideResourceRequirements
+	}
+
+	return nil
+}
+
+// GetFleetAgentTolerations returns additional tolerations for the fleet agent if it has been user defined. If not,
+// then nil is returned.
+func GetFleetAgentTolerations(cluster *v3.Cluster) []corev1.Toleration {
+	if cluster.Spec.FleetAgentDeploymentCustomization != nil &&
+		cluster.Spec.FleetAgentDeploymentCustomization.AppendTolerations != nil {
+		return cluster.Spec.FleetAgentDeploymentCustomization.AppendTolerations
+	}
+
+	return nil
+}
+
+// GetFleetAgentAffinity returns node affinity for the fleet agent if it has been user defined. If not, then the
+// default affinity is returned.
+func GetFleetAgentAffinity(cluster *v3.Cluster) *corev1.Affinity {
+	if cluster.Spec.FleetAgentDeploymentCustomization != nil &&
+		cluster.Spec.FleetAgentDeploymentCustomization.OverrideAffinity != nil {
+		return cluster.Spec.FleetAgentDeploymentCustomization.OverrideAffinity
+	}
+
+	affinity, err := unmarshalAffinity(settings.FleetAgentDefaultAffinity.Get())
+	if err != nil {
+		return settings.GetFleetAgentDefaultAffinity()
```

Comment:
Given the exported nature of this (`GetFleetAgentAffinity`) function, I think discarding the error and falling back to "Default" is not OK.

The current implementation is likely to lead to user frustration when they are trying to define their fleet-agent cluster affinity. I would much rather have this function return an error indicating that it was not possible to unmarshal the fleet agent affinity, so we can trickle this feedback upwards and ensure that the person gets the desired feedback rather than losing it.